### DISCUSSION TITLE
Where can I find documentation about timeline?

### DISCUSSION THREAD

**Discussion by zhaozheng09** [horovod/horovod#3457] [Q&A]
![image](https://user-images.githubusercontent.com/33950866/156780072-fbf3d82f-0346-4f79-917e-b07e8c5e97d4.png)
1. Is there official documentation for NEGOTIATE_ALLREDUCE WAIT_FOR_DATA  WAIT_FOR_OTHER_TENSOR_DATA MEMCPY_IN_FUSION_BUFFER NCCL_ALLREDUCE ?
2. My alltoall is too slow, where is the bottleneck?

**Comment by EnricoMi** [comment]
@tgaddair @romerojosh can you provide some insights here, please?

**Comment by maxhgerlach** [comment]
We have official documentation and it's not exactly hidden. https://horovod.readthedocs.io/en/stable/timeline.html

Your timeline shows that one NCCL_ALLTOALL is executed after the other as expected. See the greenish rectangles in your timeline screenshot.

Can we move this to discussions?

**Comment by zhaozheng09** [comment]
> We have official documentation and it's not exactly hidden. https://horovod.readthedocs.io/en/stable/timeline.html
> 
> Your timeline shows that one NCCL_ALLTOALL is executed after the other as expected. See the greenish rectangles in your timeline screenshot.
> 
> Can we move this to discussions?

OK，So alltoall is sync executable and alltoall block in queue?  How to explain this?
![image](https://user-images.githubusercontent.com/33950866/156975411-aebada65-41ec-468f-b808-cbff10d00e65.png)
thread 32 run and thread 28 blocked
thread 33 run and thread 29 blocked
and why thread 33 is blocked ?
and How can I give you this timeline file ?

**Comment by zhaozheng09** [comment]
and this:
![image](https://user-images.githubusercontent.com/33950866/157050017-6ed99a7a-5222-4756-bc7e-3a1d2cb9fb80.png)

**Comment by zhaozheng09** [comment]
![image](https://user-images.githubusercontent.com/33950866/157052391-98a4a19b-5e2b-4074-ab3e-4e4c73a386b4.png)

**Reply by zhaozheng09** [reply]
So, I think :
1. The alltoall op must be done after the gpu is idle.
2. The alltoall op in tensorflow is a parallel computation, but alltoall runs serially in mpi or horovod.

**Comment by maxhgerlach** [comment]
> The alltoall op must be done after the gpu is idle.

Hmm, I think GPU compute and communication (alltoall) should be able to overlap. Are you sure that (1) the alltoall doesn't depend on any of the results that you compute before, and that (2) all the other workers are also ready in the time frame of the highlighted region.

> The alltoall op in tensorflow is a parallel computation, but alltoall runs serially in mpi or horovod.

That's a way to put it. TensorFlow will enqueue a bunch of Horovod ops and asynchronously wait for their results (the gray HorovodAlltoall blocks in your Nsight screenshot). But typically the Horovod background thread will run just one stream of NCCL communications (you can configure this if you like), which usually makes sense to avoid competing for bandwidth.

**Reply by zhaozheng09** [reply]
< I think GPU compute and communication (alltoall) should be able to overlap. > 
1. I run model on single machine(8 gpu cards) and you can see  from nsight screenshot that a2a_op waits a long time(8-10ms)  and a2a_op in 8 cards have almost a same start time .
2. All2all op has a same start time with compute op, so I think alltoall input op is ready .
3. I adjust NCCL_CYCLE_TIME and get better performance. Now the performance of all2all doesn't seem to be the bottleneck.
so now, I want to overlap GPU compute and communication .

**Reply by maxhgerlach** [reply]
I think `HOROVOD_CYCLE_TIME` is a bit tricky to optimize unfortunately. For alltoall ops in the forward pass it makes sense to set a very low value as you have seen, but too low values may hamper tensor fusion in backward pass allreduces, which could impact training throughput. Working with explicitly grouped allreduces might help with the latter problm though.
